File: surf-api-core/surf-api-core-server/src/main/java/dev/slne/surf/surfapi/core/server/CoreInstance.java
package dev.slne.surf.surfapi.core.server;

import dev.slne.surf.surfapi.core.api.messages.Colors;
import dev.slne.surf.surfapi.core.server.listener.CoreListenerManager;
import dev.slne.surf.surfapi.core.server.util.PlayerSkinFetcher;

public class CoreInstance {

  public void onLoad() {
    // initialize classes
    PlayerSkinFetcher.class.getClassLoader();
    Colors.class.getClassLoader();
  }

  public void onEnable() {
    CoreListenerManager.INSTANCE.registerListeners();
  }

  public void onDisable() {
    CoreListenerManager.INSTANCE.unregisterListeners();
  }
}

File: surf-api-core/surf-api-core-server/src/main/kotlin/dev/slne/surf/surfapi/core/server/CoreInstance.kt
package dev.slne.surf.surfapi.core.server

import dev.slne.surf.surfapi.core.api.messages.Colors
import dev.slne.surf.surfapi.core.server.listener.CoreListenerManager
import dev.slne.surf.surfapi.core.server.util.PlayerSkinFetcher
import org.jetbrains.annotations.MustBeInvokedByOverriders

abstract class CoreInstance {

    @MustBeInvokedByOverriders
    open suspend fun bootstrap() {
        initObjects()
    }

    @MustBeInvokedByOverriders
    open suspend fun onLoad() {
    }

    @MustBeInvokedByOverriders
    open suspend fun onEnable() {
        CoreListenerManager.registerListeners()
    }

    @MustBeInvokedByOverriders
    open suspend fun onDisable() {
        CoreListenerManager.unregisterListeners()
    }

    private fun initObjects() {
        PlayerSkinFetcher
        Colors
    }
}